File: apps/api/src/travel_agent/api/http/routers/health.py
from __future__ import annotations

from typing import cast

from fastapi import APIRouter, Request, Response, status
from pydantic import BaseModel, ConfigDict

from travel_agent.bootstrap.container import Container

router = APIRouter(tags=["system"])


class CheckResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str
    status: str
    detail: str | None = None


class HealthResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    status: str
    version: str
    checks: tuple[CheckResponse, ...] = ()


def _container(request: Request) -> Container:
    return cast(Container, request.app.state.container)


@router.get("/health/live", response_model=HealthResponse)
async def live(request: Request) -> HealthResponse:
    container = _container(request)
    return HealthResponse(status="ok", version=container.settings.app_version)


@router.get("/health/ready", response_model=HealthResponse)
async def ready(request: Request, response: Response) -> HealthResponse:
    container = _container(request)
    report = await container.health_service.readiness()
    if not report.ready:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
    return HealthResponse(
        status="ready" if report.ready else "not_ready",
        version=container.settings.app_version,
        checks=tuple(
            CheckResponse(name=check.name, status=check.status, detail=check.detail)
            for check in report.checks
        ),
    )

File: apps/web/src/shared/api/health.ts
export interface HealthCheck {
  name: string
  status: string
  detail: string | null
}

export interface HealthResponse {
  status: 'ready' | 'not_ready'
  version: string
  checks: HealthCheck[]
}

export async function getReadiness(signal?: AbortSignal): Promise<HealthResponse> {
  const response = await fetch('/health/ready', {
    headers: { Accept: 'application/json' },
    signal,
  })
  if (!response.ok && response.status !== 503) {
    throw new Error(`服务返回 ${response.status}`)
  }
  return (await response.json()) as HealthResponse
}
